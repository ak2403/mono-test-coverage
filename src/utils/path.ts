import fs from "fs";

export const getDirectoriesForSource = (source: string): string[] => {
  const dirList = fs
    .readdirSync(source, { withFileTypes: true })
    .filter((dir) => dir.isDirectory())
    .map((dir) => dir.name);

  return dirList;
};

export const isFilePathExists = (filePath: string): boolean =>
  fs.existsSync(filePath);
