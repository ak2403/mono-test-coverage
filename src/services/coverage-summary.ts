import path from "path";
import fs from "fs";
import { isFilePathExists } from "../utils/path";
import { CoverageSummaryData } from "../types/services";

const DEFAULT_COVERAGE_PATH = "coverage/coverage-summary.json";

export class CoverageSummary {
  static fetch(
    sourcePath: string,
    filePath: string = DEFAULT_COVERAGE_PATH
  ): CoverageSummaryData | null {
    const coveragePath = path.join(sourcePath, filePath);

    if (!isFilePathExists(coveragePath)) {
      console.log(`No test coverage found for ${filePath}`);

      return null;
    }

    const dataBuffer = fs.readFileSync(coveragePath, {
      encoding: "utf-8",
    });

    return JSON.parse(dataBuffer) as CoverageSummaryData;
  }
}
