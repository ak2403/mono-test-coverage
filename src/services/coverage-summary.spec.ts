import path from "path";
import { CoverageSummary } from "./coverage-summary";

describe("CoverageSummary", () => {
  it("fetch the coverage summary for specified path", () => {
    const sourcePath = path.join(__dirname, "../../examples/apps/web");

    const coverageSummary = CoverageSummary.fetch(sourcePath);

    expect(coverageSummary).toMatchSnapshot();
  });
});
