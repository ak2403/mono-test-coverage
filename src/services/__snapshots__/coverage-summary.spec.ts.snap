// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoverageSummary renders 1`] = `
{
  "/Users/ak/Documents/my_project_works/mono-test-coverage/examples/apps/web/app/layout.tsx": {
    "branches": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 24,
    },
    "statements": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 24,
    },
  },
  "/Users/ak/Documents/my_project_works/mono-test-coverage/examples/apps/web/app/page.tsx": {
    "branches": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 76,
      "pct": 100,
      "skipped": 0,
      "total": 76,
    },
    "statements": {
      "covered": 76,
      "pct": 100,
      "skipped": 0,
      "total": 76,
    },
  },
  "/Users/ak/Documents/my_project_works/mono-test-coverage/examples/apps/web/components/logo.tsx": {
    "branches": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "functions": {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 7,
      "pct": 100,
      "skipped": 0,
      "total": 7,
    },
    "statements": {
      "covered": 7,
      "pct": 100,
      "skipped": 0,
      "total": 7,
    },
  },
  "/Users/ak/Documents/my_project_works/mono-test-coverage/examples/apps/web/next.config.mjs": {
    "branches": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "functions": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 4,
    },
    "statements": {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 4,
    },
  },
  "total": {
    "branches": {
      "covered": 2,
      "pct": 50,
      "skipped": 0,
      "total": 4,
    },
    "branchesTrue": {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": {
      "covered": 2,
      "pct": 50,
      "skipped": 0,
      "total": 4,
    },
    "lines": {
      "covered": 83,
      "pct": 74.77,
      "skipped": 0,
      "total": 111,
    },
    "statements": {
      "covered": 83,
      "pct": 74.77,
      "skipped": 0,
      "total": 111,
    },
  },
}
`;
